app/wasm/query_plugin.go

```diff
@@ -41,6 +44,64 @@ func CustomQuerier(osmoKeeper ViewKeeper) func(ctx sdk.Context, request json.Raw
 				return nil, sdkerrors.Wrap(err, "osmo pool state query response")
 			}
 			return bz, nil
+		} else if contractQuery.SpotPrice != nil {
+			poolId := contractQuery.SpotPrice.Swap.PoolId
+			denomIn := contractQuery.SpotPrice.Swap.DenomIn
+			denomOut := contractQuery.SpotPrice.Swap.DenomOut
+			withSwapFee := contractQuery.SpotPrice.WithSwapFee
+
+			spotPrice, err := osmoKeeper.GetSpotPrice(ctx, poolId, denomIn, denomOut, withSwapFee)
+			if err != nil {
+				return nil, sdkerrors.Wrap(err, "osmo spot price query")
+			}
+
+			res := wasm.SpotPriceResponse{Price: spotPrice.String()}
+			bz, err := json.Marshal(res)
+			if err != nil {
+				return nil, sdkerrors.Wrap(err, "osmo spot price query response")
+			}
+			return bz, nil
+		} else if contractQuery.EstimatePrice != nil {
+			sender := "" // FIXME
+			poolId := contractQuery.EstimatePrice.First.PoolId
+			denomIn := contractQuery.EstimatePrice.First.DenomIn
+			denomOut := contractQuery.EstimatePrice.First.DenomOut
+			var exactIn bool
+			var amount sdk.Int
+			var maxInOrMinOutAmount sdk.Int
+			if contractQuery.EstimatePrice.Amount.In != nil {
```

Comment:
I think just doing one basic switch here for `Amount.In` or `Amount.Out` and then sending all args unprocessed to the QueryPlugin, would make this cleaner. There are too many variables spread in 2 functions for a relatively small amount of work, I think the split complicates it.

Reply:
I wanted to separate args processing from the functional parts, but I think you are right, it's simpler the other way around. It's probably better for unit testing as well.

Reply:
In general that makes sense, if we have a larger/complex system. But since we have a relatively small amount of code that is just translating from one API to another API, it seems to add confusion here.

Reply:
Done.